src/Executor/Promise/Adapter/AmpPromiseAdapter.php

```diff
@@ -0,0 +1,149 @@
+    public function convertThenable($thenable) : Promise
+    {
+        return new Promise($thenable, $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function then(Promise $promise, ?callable $onFulfilled = null, ?callable $onRejected = null) : Promise
+    {
+        $deferred = new Deferred();
+
+        // @see https://promisesaplus.com/#point-41
+        $onResolve = static function (?Throwable $reason, $value) use ($onFulfilled, $onRejected, $deferred) : void {
+            if ($reason === null && $onFulfilled === null) {
+                $deferred->resolve($value);
+            } elseif ($reason === null) {
+                self::resolveWithCallable($deferred, $onFulfilled, $value);
+            } elseif ($onRejected === null) {
+                $deferred->fail($reason);
+            } else {
+                self::resolveWithCallable($deferred, $onRejected, $reason);
+            }
+        };
+
+        /** @var AmpPromise $adoptedPromise */
+        $adoptedPromise = $promise->adoptedPromise;
+        $adoptedPromise->onResolve($onResolve);
+
+        return new Promise($deferred->promise(), $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function create(callable $resolver) : Promise
+    {
+        $deferred = new Deferred();
+
+        $resolver(
+            static function ($value) use ($deferred) : void {
+                $deferred->resolve($value);
+            },
+            static function (Throwable $exception) use ($deferred) : void {
+                $deferred->fail($exception);
+            }
+        );
+
+        return new Promise($deferred->promise(), $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function createFulfilled($value = null) : Promise
+    {
+        $promise = new Success($value);
+
+        return new Promise($promise, $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function createRejected($reason) : Promise
+    {
+        $promise = new Failure($reason);
+
+        return new Promise($promise, $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function all(array $promisesOrValues) : Promise
+    {
+        /** @var AmpPromise[] $promises */
+        $promises = [];
+        foreach ($promisesOrValues as $key => $item) {
+            if ($item instanceof Promise) {
+                $promises[$key] = $item->adoptedPromise;
+            } elseif ($item instanceof AmpPromise) {
+                $promises[$key] = $item;
+            }
+        }
+
+        $deferred = new Deferred();
+
+        $onResolve = static function (?Throwable $reason, ?array $values) use ($promisesOrValues, $deferred) : void {
+            if ($reason) {
+                $deferred->fail($reason);
+
+                return;
+            }
+
+            $deferred->resolve(array_replace($promisesOrValues, $values));
+        };
+
+        all($promises)->onResolve($onResolve);
+
+        return new Promise($deferred->promise(), $this);
```

Comment:
If Amp promises are A+ spec-compatible then they should work with our Promise wrapper too (as [spec defines](https://promisesaplus.com/#point-46) anything having `then` method as a Promise).

In this case, this should be as simple as:

```php
public function all(array $promisesOrValues) : Promise
{
    return new Promise(Amp\Promise\all($promisesOrValues), $this);
}
```

If it doesn't accept promises other than AmpPromise instances then your loop makes sense and it should be:
```php
public function all(array $promisesOrValues) : Promise
{
    /** @var AmpPromise[] $promises */
    $promises = [];
    foreach ($promisesOrValues as $key => $item) {
        if ($item instanceof Promise) {
            $promises[$key] = $item->adoptedPromise;
        } elseif ($item instanceof AmpPromise) {
            $promises[$key] = $item;
        }
    }
    return new Promise(Amp\Promise\all($promises), $this);
}
```

But in this case I would suggest opening an issue for Amp indicating that they are not spec-compatible in this regard.

Reply:
Amphp doesn't follow either Promises A or Promises A+. See https://github.com/amphp/amp/issues/150.

Another thing is that `Amp\Promise\all()` needs to be called with promises only so we need to filter the promises and then use `array_replace` to merge the resolved values into the array of `$promisesOrValues`.

Reply:
In your code you're losing all values which were not provided as promises but as actual values..

Reply:
Yeah, I didn't realize they don't conform to any spec. This is a bit unfortunate exactly because it makes interoperability harder. But yeah, that's why we have adapters in the first place.

Reply:
They made some adjustments to make the interoperability easier such as the fact that Amphp happily accepts and adapts any ReactPHP / A+ promises you throw at it into their promises. It's just that we need to do the opposite here so it's a bit more complicated.